File: plugin/src/test/java/com/oldstats/tracking/PetTrackerTest.java
package com.oldstats.tracking;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.oldstats.api.OldStatsApiClient;
import com.oldstats.api.StatEvent;
import net.runelite.api.ChatMessageType;
import net.runelite.api.Client;
import net.runelite.api.NPC;
import net.runelite.api.Player;
import net.runelite.api.coords.WorldPoint;
import net.runelite.api.events.ChatMessage;
import net.runelite.api.events.GameTick;
import net.runelite.api.events.NpcSpawned;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

public class PetTrackerTest {
    private OldStatsApiClient apiClient;
    private Client client;
    private PetTracker tracker;

    @Before
    public void setUp() {
        apiClient = mock(OldStatsApiClient.class);
        client = mock(Client.class);
        tracker = new PetTracker(apiClient, client);

        Player localPlayer = mock(Player.class);
        when(localPlayer.getWorldLocation()).thenReturn(new WorldPoint(100, 100, 0));
        when(client.getLocalPlayer()).thenReturn(localPlayer);
    }

    private ChatMessage chatMessage(String text) {
        return new ChatMessage(null, ChatMessageType.GAMEMESSAGE, "", text, "", 0);
    }

    private NPC npcAt(String name, WorldPoint point) {
        NPC npc = mock(NPC.class);
        when(npc.getName()).thenReturn(name);
        when(npc.getWorldLocation()).thenReturn(point);
        return npc;
    }

    @Test
    public void attributesThePetToAnNpcSpawningRightNextToThePlayer() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you're being followed."));
        tracker.onNpcSpawned(new NpcSpawned(npcAt("Vorki", new WorldPoint(100, 101, 0))));

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient).enqueue(captor.capture());
        assertEquals("Vorki", ((StatEvent.PetReceived) captor.getValue()).petName);
    }

    @Test
    public void recognizesTheAlternateSneakingIntoYourBackpackMessage() {
        tracker.onChatMessage(chatMessage("You feel something weird sneaking into your backpack."));
        tracker.onNpcSpawned(new NpcSpawned(npcAt("Baby chinchompa", new WorldPoint(100, 100, 0))));

        verify(apiClient).enqueue(any(StatEvent.PetReceived.class));
    }

    @Test
    public void doesNotTreatTheNearMissWouldHaveBeenFollowedMessageAsAPet() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you would have been followed..."));
        tracker.onNpcSpawned(new NpcSpawned(npcAt("Random npc", new WorldPoint(100, 100, 0))));

        verify(apiClient, never()).enqueue(any());
    }

    @Test
    public void ignoresAnNpcSpawningFarAwayFromThePlayer() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you're being followed."));
        tracker.onNpcSpawned(new NpcSpawned(npcAt("Random npc", new WorldPoint(500, 500, 0))));

        verify(apiClient, never()).enqueue(any());
    }

    @Test
    public void timesOutToAnUnknownPetAfterTooManyTicksPassWithNoMatchingSpawn() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you're being followed."));
        for (int i = 0; i < 4; i++) {
            tracker.onGameTick(new GameTick());
        }

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient).enqueue(captor.capture());
        assertNull(((StatEvent.PetReceived) captor.getValue()).petName);
    }

    @Test
    public void anNpcSpawnWithNoPrecedingPetMessageIsIgnored() {
        tracker.onNpcSpawned(new NpcSpawned(npcAt("Random npc", new WorldPoint(100, 100, 0))));
        verify(apiClient, never()).enqueue(any());
    }
}

File: plugin/src/test/kotlin/com/oldstats/tracking/PetTrackerTest.kt
package com.oldstats.tracking

import com.oldstats.api.OldStatsApiClient
import com.oldstats.api.StatEvent
import net.runelite.api.ChatMessageType
import net.runelite.api.Client
import net.runelite.api.NPC
import net.runelite.api.Player
import net.runelite.api.coords.WorldPoint
import net.runelite.api.events.ChatMessage
import net.runelite.api.events.GameTick
import net.runelite.api.events.NpcSpawned
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class PetTrackerTest {
    private lateinit var apiClient: OldStatsApiClient
    private lateinit var client: Client
    private lateinit var tracker: PetTracker

    @Before
    fun setUp() {
        apiClient = mock()
        client = mock()
        tracker = PetTracker(apiClient, client)

        val localPlayer: Player = mock()
        whenever(localPlayer.worldLocation).thenReturn(WorldPoint(100, 100, 0))
        whenever(client.localPlayer).thenReturn(localPlayer)
    }

    private fun chatMessage(text: String): ChatMessage = ChatMessage(null, ChatMessageType.GAMEMESSAGE, "", text, "", 0)

    private fun npcAt(name: String, point: WorldPoint): NPC {
        val npc: NPC = mock()
        whenever(npc.name).thenReturn(name)
        whenever(npc.worldLocation).thenReturn(point)
        return npc
    }

    @Test
    fun `attributes the pet to an npc spawning right next to the player`() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you're being followed."))
        tracker.onNpcSpawned(NpcSpawned(npcAt("Vorki", WorldPoint(100, 101, 0))))

        val captor = argumentCaptor<StatEvent>()
        verify(apiClient).enqueue(captor.capture())
        assertEquals("Vorki", (captor.firstValue as StatEvent.PetReceived).petName)
    }

    @Test
    fun `recognizes the alternate 'sneaking into your backpack' message`() {
        tracker.onChatMessage(chatMessage("You feel something weird sneaking into your backpack."))
        tracker.onNpcSpawned(npcAt("Baby chinchompa", WorldPoint(100, 100, 0)).let { NpcSpawned(it) })

        verify(apiClient).enqueue(any<StatEvent.PetReceived>())
    }

    @Test
    fun `does not treat the near-miss 'would have been followed' message as a pet`() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you would have been followed..."))
        tracker.onNpcSpawned(NpcSpawned(npcAt("Random npc", WorldPoint(100, 100, 0))))

        verify(apiClient, never()).enqueue(any())
    }

    @Test
    fun `ignores an npc spawning far away from the player`() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you're being followed."))
        tracker.onNpcSpawned(NpcSpawned(npcAt("Random npc", WorldPoint(500, 500, 0))))

        verify(apiClient, never()).enqueue(any())
    }

    @Test
    fun `times out to an unknown pet after too many ticks pass with no matching spawn`() {
        tracker.onChatMessage(chatMessage("You have a funny feeling like you're being followed."))
        repeat(4) { tracker.onGameTick(GameTick()) }

        val captor = argumentCaptor<StatEvent>()
        verify(apiClient).enqueue(captor.capture())
        assertEquals(null, (captor.firstValue as StatEvent.PetReceived).petName)
    }

    @Test
    fun `an npc spawn with no preceding pet message is ignored`() {
        tracker.onNpcSpawned(NpcSpawned(npcAt("Random npc", WorldPoint(100, 100, 0))))
        verify(apiClient, never()).enqueue(any())
    }
}
